src/input/mod.rs

```diff
@@ -838,24 +851,32 @@ impl State {
                                         })
                                     {
                                         let layer_loc = layers.layer_geometry(layer).unwrap().loc;
-                                        if layer.can_receive_keyboard_focus()
-                                            && layer
+                                        if !layer.can_receive_keyboard_focus() {
+                                            target = KeepOld;
+                                            true
+                                        } else {
+                                            if layer
                                                 .surface_under(
                                                     relative_pos.as_logical() - layer_loc.to_f64(),
                                                     WindowSurfaceType::ALL,
                                                 )
                                                 .is_some()
-                                        {
-                                            under = Some(layer.clone().into());
-                                            true
-                                        } else {
-                                            false
+                                            {
+                                                target = Change(layer.clone().into());
+                                                true
+                                            } else {
+                                                target = Remove;
```

Comment:
I think this is somewhat confusing to read. We set the target to `Remove` implying the focus to be taken away, but since we return `false`, we never really evaluate target.

Can we please just make `target` `mut` and default initialize it with `Remove` and then skip these lines? (same goes for line 873)

Reply:
Tbh I found the entire function confusing. It's not explained what the "done" is and why it's necessary. Without knowing that, I settled on doing a mechanical translation.
I decided to drop the "mut" because it makes the choices implicit when they are not altered, making the purpose of each branch harder to figure out (need to backtrack).
I think the it can be structured better, rather than fal but I'd need a deeper explanation.

Reply:
Fair.

I don't think `done` is used to great extent any more and is mostly an artifact of previous code. It was introduced as part of [this commit](https://github.com/pop-os/cosmic-comp/commit/85edfc71418ca08bc7a0f00c634990dabb01ac6a) to break apart a never ending cascade of `if-else` statements to figure out the focus target. The previous approach had the problem of never continuing down the tree, if any layer-surface was in the way, no matter if it actually has an input_region there or even took keyboard-focus.

> I decided to drop the "mut" because it makes the choices implicit when they are not altered, making the purpose of each branch harder to figure out (need to backtrack).

Also fair, but in that case I feel like we might want to combine `done` and `target` (possibly also `under`), as the state is far too spread out at the moment imo.

> I think the it can be structured better, rather than fal but I'd need a deeper explanation.

I hope that is helpful, do you have any further questions? Do you feel like tackling this or would you prefer me doing that?

Reply:
I can do it, but I'm away for a week soon.

Reply:
> I can do it, but I'm away for a week soon.

No rush. I'll ping you, if I happen to find some time to look into this earlier than you do.